rust/trace/src/lib.rs

```diff
@@ -0,0 +1,591 @@
+#![feature(collections_range)]
+#![feature(test)]
+
+#[macro_use]
+extern crate lazy_static;
+extern crate time;
+
+#[cfg(test)]
+extern crate test;
+
+#[cfg(feature = "json_payload")]
+#[macro_use]
+extern crate serde_json;
+
+mod fixed_lifo_deque;
+
+use std::borrow::Cow;
+use std::cmp;
+
+use std::mem::size_of;
+use std::hash::{Hash, Hasher};
+use std::sync::atomic::{AtomicBool, AtomicUsize, ATOMIC_USIZE_INIT, Ordering as AtomicOrdering};
+use std::sync::Mutex;
+use fixed_lifo_deque::FixedLifoDeque;
+
+pub type StrCow = Cow<'static, str>;
+
+#[cfg(all(not(feature = "dict_payload"), not(feature = "json_payload")))]
+type TracePayloadT = StrCow;
+
+#[cfg(feature = "json_payload")]
+type TracePayloadT = serde_json::Value;
+
+#[cfg(feature = "dict_payload")]
+type TracePayloadT = std::collections::HashMap<StrCow, StrCow>;
+
+#[derive(Copy, Clone)]
+pub struct Config {
+    /* Returns the maximum number of bytes that should be used for storing trace data */
+    sample_limit_count: usize
+}
+
+impl Config {
+    fn with_limit_bytes(size: usize) -> Self {
+        Self::with_limit_count(size / size_of::<Sample>())
+    }
+
+    fn with_limit_count(limit: usize) -> Self {
+        Self {
+            sample_limit_count: limit
+        }
+    }
+
+    fn default() -> Self {
```

Comment:
All this does is make Default::default() work, right?

Reply:
It will also be callable on the implementing type, so `Config::default()` will work in both cases.

Although it probably isn't useful in this case, a major advantage of the trait version (at least that comes to mind) is that it allows for `#[derive(default)]` to work on a type that has this type as a member.